Tutorial: Get started with k6 (#1192)
grafana/k6-docs: src/data/markdown/translated-guides/en/02 Using k6/00 Tutorial/400 Reuse and re-run tests.md

      // Start iterations per `timeUnit`
      startRate: 20,
      // Start `startRate` iterations per minute
      timeUnit: "1m",
      // Pre-allocate necessary VUs.
      preAllocatedVUs: 200,
      stages: [
        { target: 20, duration: "10s" },
        { target: 40, duration: "10s" },
        { target: 40, duration: "10s" },
        { target: 60, duration: "10s" },
        { target: 60, duration: "30s" },
        { target: 0, duration: "1m" },
      ],
    },
  },
  //define thresholds
  thresholds: {
    http_req_failed: [{ threshold: "rate<0.01", abortOnFail: true }], // availability threshold for error rate
    http_req_duration: ["p(99)<1000"], // Latency threshold for percentile
  },
};

// Function to test user flow
export default function () {
  // Put visits to contact page in one group
  group("User contacts page", function () {
    // save response as variable
    const res = http.get(`${baseUrl}/contacts.php`);
    // add duration property to metric
    contactsLatency.add(res.timings.duration);
    sleep(1);
    // return to the home page, no custom metric
    http.get(`${baseUrl}/`);
    sleep(1);
  });

  // Coinflip players in another group

> 

Reviewer: ```suggestion
```

    http_req_duration: ["p(99)<1000"], // Latency threshold for percentile
  },
};
```

```json
{
  "stress_test": {
    "executor": "ramping-arrival-rate",
    "startRate": 20,
    "timeUnit": "1m",
    "preAllocatedVUs": 200,
    "stages": [
      { "target": 20, "duration": "10s" },
      { "target": 40, "duration": "10s" },
      { "target": 40, "duration": "10s" },
      { "target": 60, "duration": "10s" },
      { "target": 60, "duration": "30s" },
      { "target": 0, "duration": "1m" }
    ]
  }
}
```

</CodeGroup>

Notice the length of this final script and compare it with the script at the beginning of this page.
Though the final execution is the same, it's half the size and more readable.

Besides shortness, this modularity lets you compose scripts from many parts, or dynamically configure scripts at run time.

## Mix and match logic

With modularized configuration and logic, you can mix and match logic.
An easy way to configure this is through [environment variables](/using-k6/environment-variables).

Change `main.js` so that it:
- _By default_ runs a smoke test with 5 iterations

> - With the right enviroment variables, runs a stress test.

Reviewer: ```suggestion
- With the right enviroment variables, runs a stress test
```

To do this, follow these steps:

1. Add a new scenario file, `smoke.json`, with the following configuration

  <CodeGroup labels={["smoke.json"]} lineNumbers={[]} showCopyButton={[true]}>

  ```
  {
    "smoke": {
      "executor": "shared-iterations",
      "iterations": 5,

>       "vus": 1,

Reviewer: ```suggestion
      "vus": 1
```

>       "exec": "scenario1"

Reviewer: ```suggestion
```

this fails, we don't have an `scenario1` function.

    }
  }
  ```

  </CodeGroup>

> 1. Add conditional logic to bind `workload` to `__ENV.WORKLOAD`  if the environment variable exists, and to the content of `smoke.json` if it doesn't.

Reviewer: ```suggestion
1. Add conditional logic to bind `workload` to `__ENV.WORKLOAD`  if the environment variable exists, and to the content of `smoke.json` if it doesn't. Replace the line `const workload = JSON.parse(open("./stress.json"));` in `main.js` with:
```

  
  ```javascript
  let workload;
  if (typeof __ENV.WORKLOAD !== "undefined") {
    workload = JSON.parse(open(__ENV.WORKLOAD));
  } else {
    //if no envvar, use smoke config
    workload = JSON.parse(open("./smoke.json"));
  }
  ```
  
1. Run the script with and without the `-e` flag.

   - What happens when you run `k6 run main.js`?
   - What about `k6 run main.js -e WORKLOAD="./stress.json"`?

> If its an object, you can modularize it. Since k6 scripts are JavaScript, where essentially everything is an object, you can modularize essentially everything. 

Reviewer: ```suggestion
If it's an object, you can modularize it. Since k6 scripts are JavaScript, where everything is an object, you can modularize everything. 
```

  
## Next steps

Now you've seen examples to write tests, assert for performance, filter results, and modularize scripts.
Notice how the tests progress in complexity: from single endpoints to holistic tests, from small loads to large loads, and from single-tests to reusable modules. These progressions are typical in testing, with the next step being to automate. It might be impractical to automate a tutorial, but if you are interested,
read the [Automated performance testing](/testing-guides/automated-performance-testing/) guide, or use the [k6 GitHub action](https://github.com/grafana/k6-action).

More likely, you want to learn more about k6. [The k6 learn repository](https://github.com/grafana/k6-learn) has more detailed to practice.
Or, you can read explore the docs and try to build out your testing strategy.
The following links only scratch the surface:
- Learn how to use more [protocols](/using-k6/protocols)

> - Review the reference for the full-featured [JavaScript API](/javascript-api).

Reviewer: ```suggestion
- Review the reference for the full-featured [JavaScript API](/javascript-api)
```
